### PR TITLE
feat: include context on batch log errors

### LINKED ISSUE #53: Improve handling of "Log entry with size X  exceeds maximum size"
Hi,

Lib version: google-cloud-logging==1.15.0 

I'm experiencing lots of errors like:
```
Traceback (most recent call last):
  File "/env/lib/python3.7/site-packages/google/cloud/logging/handlers/transports/background_thread.py", line 123, in _safely_commit_batch
    batch.commit()
  File "/env/lib/python3.7/site-packages/google/cloud/logging/logger.py", line 383, in commit
    client.logging_api.write_entries(entries, **kwargs)
  File "/env/lib/python3.7/site-packages/google/cloud/logging/_gapic.py", line 126, in write_entries
    partial_success=partial_success,
  File "/env/lib/python3.7/site-packages/google/cloud/logging_v2/gapic/logging_service_v2_client.py", line 477, in write_log_entries
    request, retry=retry, timeout=timeout, metadata=metadata
  File "/env/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "/env/lib/python3.7/site-packages/google/api_core/retry.py", line 286, in retry_wrapped_func
    on_error=on_error,
  File "/env/lib/python3.7/site-packages/google/api_core/retry.py", line 184, in retry_target
    return target()
  File "/env/lib/python3.7/site-packages/google/api_core/timeout.py", line 214, in func_with_timeout
    return func(*args, **kwargs)
  File "/env/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 59, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.InvalidArgument: 400 Log entry with size 256.6K exceeds maximum size of 256.0K
```
This traceback doesn't help me identify place where the offending line was logged (I'm using python logging), so I cannot fix  it.

I can see at least 3 resolutions to this:
- allow user to specify that log message can be cropped to max size before sending to Cloud Logging. This way, if someone is ok with it, everything will get (at least partially) logged
- add to the traceback information where the offending log entry was logged (file name + line number). This way I may be able to identify and fix too long log lines - right now there's no such information.
- verify the log  message size eagerly when logging, i.e. throw exception from methods like `logging.info` when the message is too long and we won't be able to upload it (so rather than doing the verification in a background thread, do it immediately)

### PR DESCRIPTION
**Pull request by daniel-sanche**
Sometimes, a log may fail as part of a larger batch of logs (eg, due to payload size). These exceptions can be hard to track down, because batches happen asynchronously on a background thread

This PR attempts to help resolve the issue by adding a truncated copy of the failing log to the batch log exception message

```
  File "router.py", line 86, in pubsub_callback
    found_func(**kwargs)
  File "/app/snippets.py", line 81, in batch_large_log
    batch.commit()
  File "/app/python-logging/google/cloud/logging_v2/logger.py", line 470, in commit
    raise e
  File "/app/python-logging/google/cloud/logging_v2/logger.py", line 465, in commit
    client.logging_api.write_entries(entries, partial_success=True, **kwargs)
  File "/app/python-logging/google/cloud/logging_v2/_gapic.py", line 160, in write_entries
    self._gapic_api.write_log_entries(request=request)
  File "/app/python-logging/google/cloud/logging_v2/services/logging_service_v2/client.py", line 725, in write_log_entries
    metadata=metadata,
  File "/usr/local/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 154, in __call__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/google/api_core/retry.py", line 288, in retry_wrapped_func
    on_error=on_error,
  File "/usr/local/lib/python3.7/site-packages/google/api_core/retry.py", line 190, in retry_target
    return target()
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 74, in error_remapped_callable
    raise exceptions.from_grpc_error(exc) from exc
google.api_core.exceptions.InvalidArgument: 400 Log entry with size 263.8K exceeds maximum size of 256.0K: StructEntry(log_name=None, labels=None, insert_id=None, severity='DEFAULT', http_request=None, timestamp=None, resource=Resource(type='global', labels={}), trace=None, span_id=None, trace_sampled=None, source_location=None, operation=None, logger=None, payload={'message': 'simple_log', 'buffer': '0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000... [type_url: "type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors"
value: "\nA\010\007\022=\010\003\0229Log entry with size 263.8K exceeds maximum size of 256.0K"
, detail: "[ORIGINAL ERROR] generic::invalid_argument: Log entry with size 263.8K exceeds maximum size of 256.0K [google.rpc.error_details_ext] { message: \"Log entry with size 263.8K exceeds maximum size of 256.0K\" details { [type.googleapis.com/google.logging.v2.WriteLogEntriesPartialErrors] { log_entry_errors { key: 7 value { code: 3 message: \"Log entry with size 263.8K exceeds maximum size of 256.0K\" } } } } }"
]

```

Fixes: https://github.com/googleapis/python-logging/issues/53

### CODE REVIEW

File: google/cloud/logging_v2/logger.py [hunk 1]
```diff
@@ -457,6 +461,36 @@ def commit(self, *, client=None):
             kwargs["labels"] = self.logger.labels
 
         entries = [entry.to_api_repr() for entry in self.entries]
-
-        client.logging_api.write_entries(entries, **kwargs)
+        try:
+            client.logging_api.write_entries(entries, partial_success=True, **kwargs)
```
Comment (losalex) [inline google/cloud/logging_v2/logger.py:465]: As I mentioned before, seems that in https://cloud.google.com/logging/docs/reference/v2/rest/v2/entries/write the `partialSuccess` is optional

File: google/cloud/logging_v2/logger.py [hunk 2]
```diff
@@ -457,6 +461,36 @@ def commit(self, *, client=None):
...
+        Best-effort basis. If another exception occurs while processing the
+        input exception, the input will be left unmodified
+
+        Args:
+            err (~google.api_core.exceptions.InvalidArgument):
+                The original exception object
+        """
+        try:
+            # find debug info proto if in details
+            debug_info = next(x for x in err.details if isinstance(x, DebugInfo))
```
Comment (losalex) [inline google/cloud/logging_v2/logger.py:487]: Is there any public reference on that debug info structure?

**Review by losalex — COMMENTED** [review]
Lets some comments

File: google/cloud/logging_v2/logger.py (hunk 1 above)
Comment (daniel-sanche) [reply google/cloud/logging_v2/logger.py:465]: The partialSuccess field is being [modified by a different PR](https://github.com/googleapis/python-logging/pull/649/files#), so you can ignore that here

Although this improved exception message feature will only work if `partialSuccess` is `True`, which might be another reason to keep it on by default

File: google/cloud/logging_v2/logger.py (hunk 2 above)
Comment (daniel-sanche) [reply google/cloud/logging_v2/logger.py:487]: The structure is all documented as part of the [InvalidArgument exception](https://googleapis.dev/python/google-api-core/latest/exceptions.html#google.api_core.exceptions.InvalidArgument) (which was listed in the Args section above)

**Review by losalex — APPROVED** [review]

**Pull request merged**